product/util.py

```diff
@@ -154,15 +157,52 @@ def merge_products(left_product, right_product):
     return merged
 
 
-def diff_products(left_product, right_product, ignore_component_names=()):
+@dataclasses.dataclass
+class ComponentDiff:
+    cidentities_only_left: set = dataclasses.field(default_factory=set)
+    cidentities_only_right: set = dataclasses.field(default_factory=set)
+    cpairs_version_changed: list = dataclasses.field(default_factory=list)
+    # only set when new component is added/removed
+    names_only_left: set = dataclasses.field(default_factory=set)
+    names_only_right: set = dataclasses.field(default_factory=set)
+    # only set on update
+    names_version_changed: set = dataclasses.field(default_factory=set)
+
+
+# for v2 =  pass cd v2 and resolve components
+# make use of caching
+def diff_products(
+    left_product: cm.ComponentDescriptor,
+    right_product: cm.ComponentDescriptor,
+    ignore_component_names=(),
+    cache_dir=None,
+) -> ComponentDiff:
+    if type(left_product) is not cm.ComponentDescriptor:
+        raise NotImplementedError(
+            f'left product unsupported type {type(left_product)=}.'
```

Comment:
ok with me to be this verbose
in general, as this is merely a programming error (not helpful to end-users), I would tend to save some typing.

also, I think a `TypeError` would be more adequate here.

NotImplementedError fits better to indicate situations where the passed value is in general valid, but not (yet) handled. However, I think we do not plan to support different types anytime soon, do we? :-)

Reply:
no we don't :) I changed the Error type and shortened the message